File: test/pin_crypto_test.dart
import 'package:bitibridge_wallet/src/vault/pin_crypto.dart';
import 'package:flutter_test/flutter_test.dart';

const String kFrase =
    'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about';

void main() {
  group('seed cifrada pelo PIN', () {
    test('o PIN certo devolve exatamente a mesma frase', () async {
      final EncryptedSeed blob = await encryptSeedWithPin(kFrase, '123456');
      expect(await decryptSeedWithPin(blob, '123456'), kFrase);
    });

    test('PIN errado NÃO abre', () async {
      final EncryptedSeed blob = await encryptSeedWithPin(kFrase, '123456');
      expect(
        () => decryptSeedWithPin(blob, '123457'),
        throwsA(isA<WrongPinException>()),
      );
    });

    test('a frase não aparece em claro no que é gravado', () async {
      final EncryptedSeed blob = await encryptSeedWithPin(kFrase, '123456');
      final String gravado = blob.encode();
      expect(gravado.contains('abandon'), isFalse);
      expect(gravado.contains('about'), isFalse);
    });

    test('sal e nonce nunca se repetem entre gravações', () async {
      final EncryptedSeed a = await encryptSeedWithPin(kFrase, '123456');
      final EncryptedSeed b = await encryptSeedWithPin(kFrase, '123456');
      expect(a.salt, isNot(b.salt));
      expect(a.nonce, isNot(b.nonce));
      // mesma frase, mesmo PIN, cifra diferente — é o esperado
      expect(a.cipherText, isNot(b.cipherText));
    });

    test('serializa e volta sem perder nada', () async {
      final EncryptedSeed blob = await encryptSeedWithPin(kFrase, '999999');
      final EncryptedSeed volta = EncryptedSeed.decode(blob.encode());
      expect(await decryptSeedWithPin(volta, '999999'), kFrase);
    });

    test('blob corrompido é recusado', () {
      expect(() => EncryptedSeed.decode('lixo'), throwsFormatException);
    });
  });
}

File: src/lib/vault/pinCrypto.test.ts
import { describe, expect, it } from "vitest";

import {
  decodeEncryptedSeed,
  decryptSeedWithPin,
  encodeEncryptedSeed,
  encryptSeedWithPin,
  WrongPinError,
} from "./pinCrypto";

const FRASE =
  "abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about";

describe("seed cifrada pelo PIN", () => {
  it("o PIN certo devolve exatamente a mesma frase", async () => {
    const blob = await encryptSeedWithPin(FRASE, "123456");
    expect(await decryptSeedWithPin(blob, "123456")).toBe(FRASE);
  });

  it("PIN errado NÃO abre", async () => {
    const blob = await encryptSeedWithPin(FRASE, "123456");
    await expect(decryptSeedWithPin(blob, "123457")).rejects.toBeInstanceOf(WrongPinError);
  });

  it("a frase não aparece em claro no que é gravado", async () => {
    const gravado = encodeEncryptedSeed(await encryptSeedWithPin(FRASE, "123456"));
    expect(gravado).not.toContain("abandon");
    expect(gravado).not.toContain("about");
  });

  it("sal e IV nunca se repetem entre gravações", async () => {
    const a = await encryptSeedWithPin(FRASE, "123456");
    const b = await encryptSeedWithPin(FRASE, "123456");
    expect(a.salt).not.toEqual(b.salt);
    expect(a.iv).not.toEqual(b.iv);
    // mesma frase, mesmo PIN, cifra diferente — é o esperado
    expect(a.cipherText).not.toEqual(b.cipherText);
  });

  it("serializa e volta sem perder nada", async () => {
    const blob = await encryptSeedWithPin(FRASE, "999999");
    const volta = decodeEncryptedSeed(encodeEncryptedSeed(blob));
    expect(await decryptSeedWithPin(volta, "999999")).toBe(FRASE);
  });

  it("blob corrompido é recusado", () => {
    expect(() => decodeEncryptedSeed("lixo")).toThrow();
  });
});
